fix: add mypy checking + 'py.typed' files

tseaver opened the pull request.

tseaver committed:
ci: add mypy checking

tseaver committed:
chore: add py.typed file.

tseaver committed:
chore: use DEFAULT_PYTHON_VERSION to run mypy

tseaver committed:
fix: add typing for 'Client.SCOPE'

tseaver commented on google/cloud/py.typed:
```diff
@@ -0,0 +1,2 @@
+# Marker file for PEP 561.
```
I don't think so -- I actually found an issue while working on fixing stuff for `datastore` after I installed an sdist made from this branch -- `Client.SCOPE` was being inferred as `None` here, but set as `Tuple[str]` there.

But who knows?

tswast replied:
OK. I'm a still a bit wary. Can we confirm that `from google.cloud import bigquery` doesn't make mypy think the BigQuery package is publishing types with this change?

tseaver replied:
What would the test be like?  Make fresh, editable installs of `python-api-core`, `python-cloud-core`, and `python-bigquery` into a virtualenv, and then what?

tswast replied:
Then make sure something like this doesn't happen: https://issues.apache.org/jira/browse/BEAM-12975

tseaver replied:
PTAL:
- 968e2640a36987e44f4d9777d6b88fe0cf1adc9d removes the top-level `py.typed` file.
- 4c88642f3caa50cf9d7b61e27d50638f2dcbe7c3 converts all the top-level modules to packages, and adds `py.typed` within each one.

tseaver committed:
fix: cannot have 'py.typed' in the namespace package dir

tseaver committed:
chore: make top-level modules in 'google/core' into packages

FBO typing, which interprets 'py.typed' file as always-inherited in subdirs.

tseaver committed:
Merge branch 'main' into add-mypy

tswast commented on setup.py:
```diff
@@ -39,7 +39,7 @@
 
 version = {}
-with open(os.path.join(package_root, "google/cloud/version.py")) as fp:
+with open(os.path.join(package_root, "google/cloud/version/__init__.py")) as fp:
```
This might require an update to https://github.com/googleapis/release-please/blob/main/src/releasers/python.ts

tseaver replied:
Rather than that, maybe we just back out converting `google/cloud/version.py` to a package?  I don't really see anybody choking on lack of typing when importing the one attribute in that module.

tseaver replied:
Backed out in c77b9dc.

tswast replied:
Sounds good to me. It's literally just 1 string, so I'm okay with that.

tseaver committed:
chore: back out converson of 'version.py' to a package

Our 'release-please' tooling won't find it there, and nobody is likely
to a) import the module,  and then b) choke on its lack of typing.

tseaver committed:
chore: fix thinko

tswast commented:
Confirmed that it's passion on Kokoro:

```
nox > Running session mypy
nox > Creating virtual environment (virtualenv) using python3.7 in .nox/mypy
nox > python -m pip install . mypy
nox > python -m pip install types-setuptools types-requests types-mock types-protobuf
nox > mypy google tests
Success: no issues found in 18 source files
nox > Session mypy was successful.
```

https://source.cloud.google.com/results/invocations/e6d78c2c-70f3-437e-8f48-fc8996976c38/targets/cloud-devrel%2Fclient-libraries%2Fpython%2Fgoogleapis%2Fpython-cloud-core%2Fpresubmit%2Fpresubmit/log

tswast commented:
This should go in a minor release, not a patch release. I probably should have marked as `feat: ` rather than `fix: ` when merging.
